File: api/roi-calculator.py
import json
import uuid
import os
from datetime import datetime
from pymongo import MongoClient

def calculate_roi(annual_revenue, current_efficiency, target_efficiency, implementation_cost, annual_operating_cost):
    """Calculate ROI based on efficiency improvements"""
    try:
        # Calculate efficiency improvement
        efficiency_gain = target_efficiency - current_efficiency
        
        # Calculate annual savings based on efficiency gain
        # Assume efficiency translates to cost savings
        baseline_operational_cost = annual_revenue * 0.3  # 30% of revenue as operational baseline
        annual_savings = baseline_operational_cost * (efficiency_gain / 100)
        
        # Calculate net annual benefit
        net_annual_benefit = annual_savings - annual_operating_cost
        
        # Calculate total 3-year benefits
        total_benefits_3yr = net_annual_benefit * 3
        
        # Calculate ROI percentage
        roi_percentage = ((total_benefits_3yr - implementation_cost) / implementation_cost) * 100 if implementation_cost > 0 else 0
        
        # Calculate payback period in months
        if net_annual_benefit > 0:
            payback_months = (implementation_cost / net_annual_benefit) * 12
        else:
            payback_months = float('inf')
        
        return {
            'annual_savings': round(annual_savings, 2),
            'net_annual_benefit': round(net_annual_benefit, 2),
            'total_benefits_3yr': round(total_benefits_3yr, 2),
            'roi_percentage': round(roi_percentage, 1),
            'payback_months': round(payback_months, 1) if payback_months != float('inf') else None,
            'efficiency_gain': efficiency_gain
        }
    except Exception as e:
        raise ValueError(f"ROI calculation error: {str(e)}")

def handler(request):
    """ROI Calculator endpoint for Vercel"""
    
    # Handle CORS preflight
    if request.get('method') == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': {
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
                'Access-Control-Allow-Headers': 'Content-Type, Authorization',
            },
            'body': ''
        }
    
    try:
        # Parse request body
        body_str = request.get('body', '{}')
        if isinstance(body_str, bytes):
            body_str = body_str.decode('utf-8')
        body = json.loads(body_str)
        
        company_name = body.get('company_name', '').strip()
        industry = body.get('industry', '').strip()
        annual_revenue = float(body.get('annual_revenue', 0))
        current_efficiency = float(body.get('current_efficiency', 0))
        target_efficiency = float(body.get('target_efficiency', 0))
        implementation_cost = float(body.get('implementation_cost', 0))
        annual_operating_cost = float(body.get('annual_operating_cost', 0))
        email = body.get('email', '').strip().lower()
        
        # Validation
        if not email or annual_revenue <= 0:
            return {
                'statusCode': 400,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*',
                },
                'body': json.dumps({'error': 'Email and valid annual revenue are required'})
            }
        
        if '@' not in email or '.' not in email:
            return {
                'statusCode': 400,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*',
                },
                'body': json.dumps({'error': 'Invalid email format'})
            }
        
        if current_efficiency >= target_efficiency:
            return {
                'statusCode': 400,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*',
                },
                'body': json.dumps({'error': 'Target efficiency must be higher than current efficiency'})
            }
        
        # Calculate ROI
        roi_results = calculate_roi(
            annual_revenue, current_efficiency, target_efficiency,
            implementation_cost, annual_operating_cost
        )
        
        # Database connection
        mongo_url = os.environ.get('MONGO_URL', 'mongodb://localhost:27017')
        db_name = os.environ.get('DB_NAME', 'orgainse_consulting')
        
        client = MongoClient(mongo_url)
        db = client[db_name]
        
        # Create ROI calculation record
        roi_calculation = {
            'id': str(uuid.uuid4()),
            'company_name': company_name,
            'industry': industry,
            'email': email,
            'inputs': {
                'annual_revenue': annual_revenue,
                'current_efficiency': current_efficiency,
                'target_efficiency': target_efficiency,
                'implementation_cost': implementation_cost,
                'annual_operating_cost': annual_operating_cost
            },
            'results': roi_results,
            'calculated_at': datetime.utcnow(),
            'source': 'website_roi_calculator'
        }
        
        # Save to database
        db.roi_calculations.insert_one(roi_calculation)
        client.close()
        
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps({
                'calculation_id': roi_calculation['id'],
                'results': roi_results,
                'message': 'ROI calculation completed successfully'
            }, default=str)
        }
        
    except (ValueError, TypeError) as e:
        return {
            'statusCode': 400,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps({'error': f'Invalid input data: {str(e)}'})
        }
    except json.JSONDecodeError:
        return {
            'statusCode': 400,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps({'error': 'Invalid JSON data'})
        }
    except Exception as e:
        return {
            'statusCode': 500,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
            },
            'body': json.dumps({'error': f'Internal server error: {str(e)}'})
        }

File: api/roi-calculator.js
import { MongoClient } from 'mongodb';
import { securityHeaders, rateLimit, sanitizeInput, validateEmail, validateRequestSize } from './middleware/security.js';

// Generate UUID-like ID without external dependency
function generateId() {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
    const r = Math.random() * 16 | 0;
    const v = c == 'x' ? r : (r & 0x3 | 0x8);
    return v.toString(16);
  });
}

// MongoDB connection
let cachedClient = null;
let cachedDb = null;

async function connectToDatabase() {
  if (cachedClient && cachedDb) {
    return { client: cachedClient, db: cachedDb };
  }

  const client = new MongoClient(process.env.MONGO_URL);
  await client.connect();
  
  const db = client.db(process.env.DB_NAME || 'orgainse-consulting');
  
  cachedClient = client;
  cachedDb = db;
  
  return { client, db };
}

// Regional pricing configuration
const REGION_CONFIG = {
  US: { pppMultiplier: 1.0, currency: 'USD', symbol: '$' },
  IN: { pppMultiplier: 5.5, currency: 'INR', symbol: '₹' },
  GB: { pppMultiplier: 0.85, currency: 'GBP', symbol: '£' },
  AE: { pppMultiplier: 0.75, currency: 'AED', symbol: 'AED' },
  AU: { pppMultiplier: 0.90, currency: 'AUD', symbol: 'A$' },
  NZ: { pppMultiplier: 0.85, currency: 'NZD', symbol: 'NZ$' },
  ZA: { pppMultiplier: 0.35, currency: 'ZAR', symbol: 'R' }
};

// Calculate employee count from range
function getEmployeeCountFromRange(range) {
  const ranges = {
    '1-10': 5,
    '11-50': 30,
    '51-200': 125,
    '201-500': 350,
    '500+': 750
  };
  return ranges[range] || 100;
}

// Calculate base implementation cost based on company size
function calculateBaseCost(employeeCount, timeline) {
  let baseCost = 25000; // Base cost in USD
  
  // Scale by employee count
  if (employeeCount <= 10) {
    baseCost = 15000;
  } else if (employeeCount <= 50) {
    baseCost = 25000;
  } else if (employeeCount <= 200) {
    baseCost = 45000;
  } else if (employeeCount <= 500) {
    baseCost = 75000;
  } else {
    baseCost = 120000;
  }
  
  // Adjust by timeline (faster = more expensive)
  const timelineMultipliers = {
    '1-3 months': 1.3,
    '3-6 months': 1.0,
    '6-12 months': 0.85,
    '12+ months': 0.75
  };
  
  const multiplier = timelineMultipliers[timeline] || 1.0;
  return baseCost * multiplier;
}

// Calculate ROI metrics
function calculateROI(inputs, userRegion = 'US') {
  // Parse inputs
  const annualRevenue = parseFloat(inputs.annual_revenue) || 0;
  const currentPMCosts = (parseFloat(inputs.current_pm_costs) || 0) * 12; // Annual
  const employeeCount = getEmployeeCountFromRange(inputs.employee_count || '11-50');
  
  // AI implementation benefits (based on industry research)
  const efficiencyGain = 0.25; // 25% faster delivery
  const costReduction = 0.45; // 45% cost reduction in operations
  const revenueBoost = 0.15; // 15% revenue increase from faster delivery
  
  // Calculate savings
  const operationalSavings = currentPMCosts * costReduction;
  const revenueIncrease = annualRevenue * revenueBoost;
  const totalAnnualBenefit = operationalSavings + revenueIncrease;
  
  // Calculate implementation costs (region-adjusted)
  const baseCost = calculateBaseCost(employeeCount, inputs.implementation_timeline);
  const regionalConfig = REGION_CONFIG[userRegion] || REGION_CONFIG.US;
  const regionalCost = baseCost * regionalConfig.pppMultiplier;
  
  // ROI calculations
  const roi = totalAnnualBenefit > 0 ? ((totalAnnualBenefit - regionalCost) / regionalCost) * 100 : 0;
  const paybackPeriod = totalAnnualBenefit > 0 ? Math.ceil(regionalCost / (totalAnnualBenefit / 12)) : 0;
  
  return {
    potential_savings: Math.round(totalAnnualBenefit),
    roi_percentage: Math.round(roi),
    payback_period: Math.max(1, paybackPeriod), // Minimum 1 month
    implementation_cost: Math.round(regionalCost),
    monthly_savings: Math.round(totalAnnualBenefit / 12),
    regional_currency: regionalConfig.currency,
    currency_symbol: regionalConfig.symbol,
    efficiency_gain_percent: Math.round(efficiencyGain * 100),
    cost_reduction_percent: Math.round(costReduction * 100),
    revenue_boost_percent: Math.round(revenueBoost * 100)
  };
}

export default async function handler(req, res) {
  // Apply security headers
  securityHeaders(req, res);
  
  // Handle preflight requests
  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }
  
  // Only allow POST requests
  if (req.method !== 'POST') {
    return res.status(405).json({ 
      error: 'Method not allowed',
      message: 'Only POST requests are supported for ROI calculations'
    });
  }
  
  try {
    // Apply rate limiting
    if (!rateLimit(req, res, { max: 50, windowMs: 15 * 60 * 1000 })) {
      return; // Rate limit exceeded
    }

    // Validate request size
    if (!validateRequestSize(req, res, 10 * 1024)) { // 10KB limit
      return;
    }

    const { db } = await connectToDatabase();
    
    // Sanitize input data
    const sanitizedBody = sanitizeInput(req.body);
    const { 
      company_name, 
      email, 
      annual_revenue, 
      employee_count, 
      current_pm_costs,
      tech_budget,
      implementation_timeline,
      user_region = 'US'
    } = sanitizedBody;
    
    if (!company_name || !email) {
      return res.status(400).json({
        error: 'Validation failed',
        message: 'Company name and email are required'
      });
    }

    // Validate email format
    if (!validateEmail(email)) {
      return res.status(400).json({ 
        error: 'Invalid email format' 
      });
    }
    
    if (!annual_revenue || !current_pm_costs) {
      return res.status(400).json({
        error: 'Validation failed',
        message: 'Annual revenue and current PM costs are required for ROI calculation'
      });
    }
    
    // Calculate ROI metrics
    const calculatedMetrics = calculateROI(req.body, user_region);
    
    // Prepare ROI calculation data
    const roiData = {
      calculation_id: generateId(),
      business_inputs: {
        company_name,
        email,
        annual_revenue: parseFloat(annual_revenue),
        employee_count,
        current_pm_costs: parseFloat(current_pm_costs),
        tech_budget: parseFloat(tech_budget) || null,
        implementation_timeline,
        user_region
      },
      calculated_metrics: calculatedMetrics,
      submitted_at: new Date().toISOString(),
      leadType: "ROI Calculator",
      source: "roi_calculator_tool"
    };
    
    // Insert into MongoDB
    const result = await db.collection('roi_calculator_leads').insertOne(roiData);
    
    // Return response
    return res.status(200).json({
      success: true,
      calculation_id: roiData.calculation_id,
      company_name: company_name,
      potential_savings: calculatedMetrics.potential_savings,
      roi_percentage: calculatedMetrics.roi_percentage,
      payback_period: calculatedMetrics.payback_period,
      implementation_cost: calculatedMetrics.implementation_cost,
      monthly_savings: calculatedMetrics.monthly_savings,
      currency: calculatedMetrics.regional_currency,
      currency_symbol: calculatedMetrics.currency_symbol,
      message: "ROI calculation completed successfully",
      timestamp: roiData.submitted_at,
      detailed_metrics: {
        efficiency_gain: `${calculatedMetrics.efficiency_gain_percent}%`,
        cost_reduction: `${calculatedMetrics.cost_reduction_percent}%`,
        revenue_boost: `${calculatedMetrics.revenue_boost_percent}%`
      }
    });
    
  } catch (error) {
    console.error('ROI Calculator API Error:', error);
    return res.status(500).json({
      error: 'Internal server error',
      message: 'Failed to process ROI calculation. Please try again.'
    });
  }
}